### PR TITLE
cleanup transactional file implementation

### PR DESCRIPTION
**Pull request by tobymao**

### CODE REVIEW

File: sqlmesh/utils/__init__.py [hunk 1]
```diff
@@ -45,6 +48,19 @@ def random_id() -> str:
     return uuid.uuid4().hex
 
 
+def time_uuid(k: int = 4) -> str:
```
Comment (izeigerman) [inline sqlmesh/utils/__init__.py:51]: this is not uuid though

Comment (izeigerman) [reply sqlmesh/utils/__init__.py:51]: why do we need this exactly?

File: sqlmesh/utils/transactional_file.py [hunk 2]
```diff
@@ -26,48 +26,37 @@ def __init__(self, path: str, fs: fsspec.AbstractFileSystem) -> None:
...
-        for lock in (
-            lock_info
-            for lock_info in self._fs.glob(
-                self.lock_prefix + "*", refresh=True, detail=True
-            ).values()
-            if os.path.basename(lock_info["name"]).startswith(os.path.basename(self.lock_prefix))
-        ):
+        now = datetime.now()
+
+        for lock in self._fs.glob(self.lock_prefix + ".*", refresh=True, detail=True).values():
```
Comment (izeigerman) [inline sqlmesh/utils/transactional_file.py:44]: why `detail=True`?

File: sqlmesh/utils/__init__.py (hunk 1 above)
Comment (izeigerman) [reply sqlmesh/utils/__init__.py:51]: ok, I understand why we need this, but we shouldn't call this uuid I think

Comment (tobymao) [reply sqlmesh/utils/__init__.py:51]: it is due to the k

it's possible that two files are created at the exact same time, in that case you need to sort by this as a backup

File: sqlmesh/utils/transactional_file.py (hunk 2 above)
Comment (tobymao) [reply sqlmesh/utils/transactional_file.py:44]: i have no idea, this was @z3z1ma's code

File: sqlmesh/utils/__init__.py (hunk 1 above)
Comment (tobymao) [reply sqlmesh/utils/__init__.py:51]: what do you want to call it

File: sqlmesh/utils/transactional_file.py [hunk 3]
```diff
@@ -102,27 +93,28 @@ def acquire_lock(self, blocking: bool = True, timeout: float = -1) -> bool:
         not (for example if the timeout expired).
         """
         if self._is_locked:
-            # Make this idempotent for a single instance.
             return True
-        wait_time = 0.0
+
         self._fs.touch(self.lock_path)
         locks = self._sync_locks()
         active_lock = min(locks, key=locks.get)  # type: ignore
```
Comment (izeigerman) [inline sqlmesh/utils/transactional_file.py:100]: This took me a minute to parse. Can we just do `min(locks.values())` instead?

File: sqlmesh/utils/transactional_file.py [hunk 4]
```diff
@@ -102,27 +93,28 @@ def acquire_lock(self, blocking: bool = True, timeout: float = -1) -> bool:
...
                 self._fs.rm(self.lock_path)
                 return False
-            time.sleep(POLLING_INTERVAL)
-            wait_time += POLLING_INTERVAL
+
+            time.sleep(random.random() + POLLING_INTERVAL)
             locks = self._sync_locks()
-            if not locks:
-                break
+            if self.lock_path not in locks:
```
Comment (izeigerman) [inline sqlmesh/utils/transactional_file.py:110]: So theoretically if the timeout is set to -1 it's still bounded by the lock file TTL. This is a non-obvious behavior.

File: sqlmesh/utils/transactional_file.py [hunk 5]
```diff
@@ -102,27 +93,28 @@ def acquire_lock(self, blocking: bool = True, timeout: float = -1) -> bool:
...
+                return False
+
             active_lock = min(locks, key=locks.get)  # type: ignore
```
Comment (izeigerman) [inline sqlmesh/utils/transactional_file.py:113]: Ditto: `min(locks.values())`

File: sqlmesh/utils/transactional_file.py (hunk 4 above)
Comment (tobymao) [reply sqlmesh/utils/transactional_file.py:110]: should i recreate the lockfile in this case?

File: sqlmesh/utils/transactional_file.py (hunk 2 above)
Comment (tobymao) [reply sqlmesh/utils/transactional_file.py:44]: https://filesystem-spec.readthedocs.io/en/latest/api.html?highlight=glob#fsspec.spec.AbstractFileSystem.ls i think you need detail for the timestamps

File: sqlmesh/utils/transactional_file.py (hunk 4 above)
Comment (izeigerman) [reply sqlmesh/utils/transactional_file.py:110]: I believe we should recreate the file instead or do something to it to update its mtime

File: sqlmesh/utils/transactional_file.py [hunk 6]
```diff
@@ -136,7 +128,9 @@ def release_lock(self) -> None:
             self._original_contents = None
 
     @contextmanager
-    def read_with_lock(self, timeout: float = 61.0) -> t.Iterator[t.Optional[bytes]]:
+    def read_with_lock(
```
Comment (izeigerman) [inline sqlmesh/utils/transactional_file.py:131]: I don't like this API. I think the context manager API should be `def with_lock()` to safely acquire and release the lock, and then we can do reading / writing / whatever else inside.

File: sqlmesh/utils/__init__.py (hunk 1 above)
Comment (izeigerman) [reply sqlmesh/utils/__init__.py:51]: I think I understand why we need this, but we shouldn't be calling this uuid.

Comment (izeigerman) [reply sqlmesh/utils/__init__.py:51]: maybe just `lock_id`? I feel like this logic is quite specific to the lock implementation. We can move it accordingly too

**Pull request merged**